Updated docs for the coverage model (some reorganization + new ideas)

mbabadi opened the pull request:
Some organizational changes in the doc as well as new sections. The major addition is a regularization scheme for avoiding picking up events as noise. The regularization is expected to significantly increase sensitivity when dealing with small cohorts. @davidbenjamin, please kindly review. @samuelklee, please kindly comment.

davidbenjamin commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -670,125 +712,170 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
 \end{align}
 %
-where $\vz \in \mathbb{R}^D$ is a low-dimensional latent vector of laboratory conditions, $\vW \in \mathbb{R}^{T \times D}$ is a linear map from latent space to target space, $\vm \in \mathbb{R}^T$ is the vector of mean biases, and $\vPsi \in \mathbb{R}^{T \times T}$ is a diagonal matrix of residual variance not explained by the latent features.  We approximate the Poisson as a Gaussian centered  ...
+where $\vz \in \mathbb{R}^D$ is a low-dimensional latent vector of laboratory conditions, $\vW \in \mathbb{R}^{T \times D}$ is a linear map from latent space to target space, $\vm \in \mathbb{R}^T$ is the vector of mean biases, and $\vPsi \in \mathbb{R}^{T \times T}$ is a diagonal matrix of residual variance not explained by the latent features. We approximate the Poisson as a Gaussian (an excell ...
```
The comment `an excellent approximation for $n_{st} \gtrsim 1000$` implies that the model only works for very high coverage.

mbabadi replied:
I agree. I removed it since we assess this approximation later anyway.

davidbenjamin commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -670,125 +712,170 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
+P(\vn, \vz | \vc, \vW, \vm, \vPsi) \propto \prod_s \norm(\vz_s | \vzero, \vI) \prod_{t | \MM_{st} \neq 0} \norm(m_{st} | (\vW \vz_s)_t + m_t, \Psi_t + \Sigma_{st})
 \label{coverage_likelihood}
 \end{equation}
+Save for the presence of a sample-dependent bias uncertainy $\Sigma_{st}$, our learning objective is essentially the factor analysis problem. The MLE for parameters $(\vPsi, \vW, \vm)$ can be obtained via the EM algorithm. The presence of $\Sigma_{st}$, in particular its dependence on $s$, introduces undesirable complexities which we wish to avoid. To gain more insight about the role of $\Sigma_{ ...
```
typo: `mrginalizing` -> `marginalize`

davidbenjamin replied:
using *the* Woodbury identity

samuelklee replied:
typo: uncertainy -> uncertainty

mbabadi replied:
fixed

davidbenjamin commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -670,125 +712,170 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
 \end{equation}
-The approximation that led to this can be interpreted as the replacement $\vSigma_s \rightarrow \bar{\vSigma} \equiv \frac{1}{S} \sum_s \vSigma_s$ and we will consider it again below when we discuss EM updates of $\vW$ and $\vPsi$.
+%
+To simplify the discussion, let us assume that all samples share the same targets, i.e. $\vM_s = \mathbb{I}$.
```
Either write `replacing $\Sigma_{st}$  with . . .$\Sigma_t$` or `replacing $\bar{\vSigma}_s$  with . . .$\bar{\vSigma}$`

mbabadi replied:
I actually got rid of this part since we don't use the approximation anymore.

davidbenjamin commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -670,125 +712,170 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
 %
-
-For the M steps, we must maximize with respect to $E \left[ \sum_s \ln {\rm N}(\vm_s | \vW \vz_s + \vm, \vPsi + \vSigma_s) \right]$, which reduces to
+In the M step, we calculate the expectation value of the complete-data log likelihood with respect to the posperior estimate of $\vz_s$. Save for terms independent of the model parameters, the result is:
```
`posperior` -> `posterior`

mbabadi replied:
fixed

davidbenjamin commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -670,125 +712,170 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
+One may start iterations using the result of Scheme 1 as the initial guess and continue until convergence.\\
 
-In the M step equation for $\vW$ we incorporate an automatic relevance determination (ARD) prior:
+\noindent{\bf Remark:} In practice, when sample-to-sample variables of $\Sigma$ was large (e.g. read depths varying from 50 to 1000 randomly), we noticed that the best approach was to use Brent root finding for each target. On average, 10 function calls yields to solution within a $10^{-6}$ tolerance. The Newton's method required approximately 20 evaluations of $\alpha_t$ and $\beta_t$ to converg ...
```
How well does naive averaging (which, while unprincipled, is probably less risky than a first-order expansion when depths vary a lot) perform versus Brent and Newton in this case?

samuelklee replied:
convergance -> convergence

samuelklee replied:
sample-to-sample variables -> sample-to-sample variance

yields to solution -> yields a solution

The Newton's method -> Newton's method

Interesting that starting at $\Psi_t = 0$ was more robust, but I'm not sure if I understand why---could you expand?

mbabadi replied:
fixed.

Newton's method requires certain difficult-to-a-priori-assess conditions to be met for the initial guess to converge. I have no idea why $\Psi_t = 0$ is good starting guess.

davidbenjamin commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -670,125 +712,170 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
+\end{equation}
+This solution is accurate to linear order in deviations of $\Sigma_{st}$ about $\bar{\Sigma}_t$.\\
+
+\noindent {\bf (Scheme 2) Newton iterations:} The complexity of solving Eq.~\eqref{eq:psi_stationarity} numerically is not too high given that the Hessian matrix is diagonal. Expanding $\mathcal{L}$ about $\vPsi_0$, we find:
```
I don't think you need to mention that the Hessian is diagonal because you have already cast it as independent optimizations for each t.

davidbenjamin commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -670,125 +712,170 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
 \end{align}
 %
-If $\alpha_\mu \rightarrow \infty$ latent feature $\mu$ is effectively turned off.  Thus we can initially choose a liberal estimate of $D$ and the model will automatically become more parsimonious.  The M step log likelihood times the ARD prior depend on the $t$-th row $\vW_{t \cdot}$ of $\vW$ as
+If $\alpha_\mu \rightarrow \infty$, latent feature $\mu$ is effectively turned off. Thus we can initially choose a liberal estimate of $D$ and the model will automatically become more parsimonious. The M step log likelihood times the ARD prior depend on the $t$-th row $\vW_{t \cdot}$ of $\vW$ as:
```
the latent feature

samuelklee replied:
I find the notation $\vW_{t \cdot}$ a bit weird.  The $\cdot$ is sometimes hard to see and in one case ambiguous with a period at the end of an equation.  Why not just use lowercase w as in Bishop?

mbabadi replied:
I agree -- I got rid of $\vW_{t \cdot}$ in favor of explicitly indexed matrices.

davidbenjamin commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -670,125 +712,170 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
-In the approximation $\vPsi_s \rightarrow \bar{\vPsi}$ this formula is unchanged but $\vQ_t$ is $S$ times as fast to calculate.  The other M steps and the E step are not affected by the ARD prior.
-
-To determine $\alpha_\mu$, we re-exponentiate Equation $\ref{ARD_likelihood}$\footnote{This is the distribution on $\vW$ that we would obtain from a mean-field variational factorization $q(\vz) q(\vW)$.} and integrate out $\vW$ to obtain the evidence for $\vA$:
+In the approximation $\vSigma_s \rightarrow \bar{\vSigma}$, this formula is unchanged but $\vQ_t$ is $S$ times as fast to calculate. The other M steps and the E step are not affected by the ARD prior. To determine $\alpha_\mu$, we re-exponentiate Eq.~$\eqref{ARD_likelihood}$\footnote{This is the distribution on $\vW$ that we would obtain from a mean-field variational factorization $q(\vz) q(\vW)$ ...
```
\eqref{ARD_likelihood} gave something weird in the pdf.

mbabadi replied:
fixed.

davidbenjamin commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -670,125 +712,170 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
 \end{equation}
 %
-The maximum a posterior result for $\vW_{t \cdot}$ is
+The maximum a posterior result for $\vW_{t \cdot}$ is:
```
a posteriori

mbabadi replied:
fixed

davidbenjamin commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -829,44 +916,99 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
-$\vG$ is $O(D^2 T)$.  $E[\vz_s \vz_s^T]$ is $O(D^2 S)$.  $\vG_s \vW^T \vPsi_s^{-1}$, which is common to all $E[\vz_s]$, is $O(D^2 T)$ after which $E[\vz_s]$ is $O(D T S)$.  $E[\vz_s \vz_s^T]$ is $O(D^2 S)$.  The E step overall is $O(D T S)$.
-
-\item $\vm = \frac{1}{S} \sum_s \vm_s$ is a one-time cost of $O(T S)$.
+The leading cost is a few terms of $O(D^2 T S)$ flops, each with small prefactors, per iteration. Assuming a total prefactor of $10$ and $T = 2 \times 10^5$, $D = 10$, $S = 500$ a full EM iteration costs $10^{11}$ flops in exact mode. On a single 1 GHz core ($10^9$ flops per second) this comes out to roughly 100 or 10 seconds.\\
```
With the new way of getting \bar{\vSigma} there is no longer a fast mode so the 10 second version no longer exists.

mbabadi replied:
fixed.

davidbenjamin commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -829,44 +916,99 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
+\begin{equation}
+P(\vn | \vc, \vW, \vm, \vPsi) \propto \prod_t \exp \left[ -\frac{1}{2} \, M_{st} \Psi_{st}^{-1} \big( \ln \left( n_{st} / (d_s \PP_{st}) \right) - \ln(c_{st}) - m_t - \left(\vW \EE[\vz_{s}]\right)_t \big)^2 \right],
+\end{equation}
+We have only kept the factors that depends on $\vc$ in the above likelihood. Note that this is factorized into independent likelihood terms for each target and is thus suitable for the emission model of an HMM.  This likelihood is not Gaussian in $\vc$, but it does not need to be for the Viterbi and forward-backward algorithms. Also, note that when $c_{st} = 0$ is the most likely solution, this m ...
```
factors that depend

mbabadi replied:
fixed

davidbenjamin commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -829,44 +916,99 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
 
-\item $ \vW_{t \cdot} = (\vA + \vQ)^{-1} \vv_t$ is $O(D^2 T)$.
+\subsubsection{PCA and the curse of small samples}
+PCA-like denoising approaches minimize the {\em total variance} by learning and subtracting the contribution of the underlying latent features from the data. In practice, this objective is acheived using either the maximum variance principle (usual PCA) or the maximum likelihood principle on a linear-Gaussian model as explained earlier. In either method, when the number of samples largely exceeds ...
```
achieved

samuelklee replied:
bsample -> sample

mbabadi replied:
fixed

davidbenjamin commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -829,44 +916,99 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
+PCA-like denoising approaches minimize the {\em total variance} by learning and subtracting the contribution of the underlying latent features from the data. In practice, this objective is acheived using either the maximum variance principle (usual PCA) or the maximum likelihood principle on a linear-Gaussian model as explained earlier. In either method, when the number of samples largely exceeds ...
 
-\item $\ave{W_{t\mu}^2} = W_{t\mu}^2 + (\vA + \vQ)^{-1}_{\mu\mu}$ is $O(D T)$.
+Let us assume that we have an oracle for the first $N$ major latent features, and that we have already subtracted the contribution arising from these features. Let $\sigma_\ell^2$ further be the variance associated to next leading latent feature such that subtracting it reduces the total variance by $N\sigma_\ell^2$. Now, if one of the samples has an individual leftover variance of magnitude $\si ...
```
`Let $\sigma_\ell^2$ further be the variance associated to next leading latent feature` -> `Let $\sigma_\ell^2$ be the variance associated with the next leading latent feature`

davidbenjamin replied:
I think you mean `$N\sigma_\ell^2$` -> `$S\sigma_\ell^2$`, right?

samuelklee replied:
Yes, and I think it would also be good to reiterate here the meaning of $S$.  As it reads now, it's only implied that $N$ is supposed to be the number of samples and that we just happen to know the same number of latent features.

mbabadi replied:
Yes, $N$ -> $S$ indeed. I also reiterated the meaning of $S$. Also, I got rid of $N$ altogether.

davidbenjamin commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -829,44 +916,99 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
 
-\item $\alpha_\mu = T / \sum_t \ave{W_{t\mu}^2}$ is $O(D T)$.
-\end{itemize}
+\noindent {\bf Why does it matter? ---} We have no theoretical guarantee that the MLE problem for $(\vW, \vm, \vPsi, \vc)$ is convex. In all likelihood, if one of the samples has a large germline CNV event, it may be picked up as a principal component and get washed out such that the MLE for $\vc$ fails to call that CNV event. It is possible that the likelihood function has numerous local maxima  ...
```
"get washed out" is not clear.  You could say "denoised", or nothing at all, or "interpreted as experimental noise" etc.

davidbenjamin replied:
"numerous such local maxima"

mbabadi replied:
fixed, fixed.

davidbenjamin commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -829,44 +916,99 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
+\noindent {\bf (Idea 1) Blind source separation ---} One remedy is to use a blind source separation approach, such as independent component analysis (ICA), to separate the signal from the noise as a first step, followed by learning the latent features of the noise using PCA. In ICA-like methods, one decomposes the signal into additive subcomponents and minimizes the mutual information between the ...
 
-We can apply the parameters learned from the panel of normals to single-sample calling, which requires the likelihood as a function of the copy numbers $\vc$.  Applying the same E step as above, the likelihood is
+\noindent {\bf (Idea 2) Imposing context-specific constraints on the structure of variations ---} Fortunately, we have some idea about the spatial structure of the CNV events: they are tandem amplification or attenuation of the read count over several targets. In the absence of noise, we expect the frequency spectrum of the CNV signal (as obtained by taking a Fourier transform of $m_{st}$ in $t$) ...
```
instead of "tandem" I would just say "over several consecutive targets"

samuelklee replied:
non-Gaussanity -> non-Gaussianity

mbabadi replied:
fixed.

davidbenjamin commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -829,44 +916,99 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
+Here, $k^* = \min(k, T - 1 - k)$, $k_l \sim \lfloor T / \ell_\mathrm{max} \rfloor$ and $k_l \sim \lfloor T / \ell_\mathrm{min} \rfloor$, where $\ell_\mathrm{min}$ and $\ell_\mathrm{max}$ denote roughly the minimum and maximum length of the CNV events in the units of targets. The filtered spectral power of the noise in sample $s$ is given as:
 \begin{equation}
-P({\rm data} | \vc) \propto \prod_t \exp \left[ -\frac{1}{2} \left(  (\vm_t - \vm_{st})^2 +2 (\vm_t - \vm_{st}) \vW_{t} E[\vz_{s}] \right)\Psi_{st}^{-1} \right],
+\kappa_s \equiv \sum_{k=0}^{T-1} \tilde{f}(k) \left|\mathfrak{F}\left[\vW \vz_s\right]_k\right|^2 = \frac{1}{T}\sum_{t,t'=0}^{T-1}F_{tt'} \, W_{t\mu} \, W_{t'\nu} \, z_{s\mu} \, z_{s\nu},
```
The \mathfrac letter is hard to read.

samuelklee replied:
Maybe use \mathcal?

mbabadi replied:
I used $\mathrm{FFT}$.

davidbenjamin commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -829,44 +916,99 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
+As one would expect, the regularizer mixes different targets such that $t$ is no longer a mere label in the stationarity condition for $\vW$. The direct solution to Eq.~\eqref{eq:W_eq_reg} is impractical since it involves inverting a matrix of size $DT \times DT$.
 
-We can easily integrate sample-specific GC bias into this model.  Let $f_s({\rm GC})$ be the GC bias of GC content ${\rm GC}$ for sample $s$.  Then this enters into the model as an additional multiplier to the Poisson parameter.  That is, we replace $d_s c_{st} \rightarrow d_s c_{st} f_s({\rm GC}_t)$, which affects the model learning and inference only via the definition of $m_{st}$.  We can iter ...
+Fortunately, the linear operator in question, $\vQ + \lambda \vZ \otimes \vF$, is the sum of two sparse operators: $\vQ$ is diagonal in the target space, and $\mathbf{A} \equiv \vZ \otimes \vF$ is diagonal in Fourier space (both are dense in the latent space, which has a low dimensionality anyway). Eq.~\eqref{eq:W_eq_reg} can be solved very efficiently using preconditioned iterative Krylov space  ...
```
It can't hurt to add a parenthetical about the direct product that \vZ acts in latent space and \vF acts in target space.

mbabadi replied:
agreed.

davidbenjamin commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -829,44 +916,99 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
+\noindent {\em Choice of $\lambda$ ---} The regularizer ``kicks in'' when $\lambda \sim \Psi^{-1}$, as it can be inferred directly from Eq.~\eqref{eq:W_eq_reg}. One may initially choose $\lambda \sim 1000\,\Psi^{-1}$ and progressively relax it as CNV calls stabilize.
+
+\subsubsection{Results}
+In this section, we present the result of the algorithm on synthetic converage data where the ground truth is known (this section must be eventually suplemmented with real data). We synthesize the data according to Eq.~\eqref{coverage_model} along with random duplication events of varying lengths. We choose $T=4000$ targets, $D=10$ true latent variables, $S=100$ samples, mean read depth $d$ unifo ...
```
$d$ uniformly sampled

samuelklee replied:
converage -> coverage

samuelklee replied:
suplemmented -> supplemented

samuelklee replied:
I'd say "uniformly in -> uniformly sampled from" is fine.

mbabadi replied:
fixed all.

davidbenjamin commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -829,44 +916,99 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
+
+Figs.~\ref{fig:comp_random} and~\ref{fig:comp_correlated} compares PCA denoising against our probablistic model with different features turned on/off (ARD, CNV event regularization) for random and correlated events, respectively. It is clearly observed that the regularized model retains all of the events even when the number of latent features is chosen larger than that of the ground truth.
+
+\begin{figure}
```
Figures 5 and 6 need to be bigger.  And it would be nice to force them to be together, too.

samuelklee replied:
probablistic -> probabilistic

samuelklee replied:
is chosen -> chosen is

samuelklee replied:
I agree about making the figures larger---right now, the target tick labels are running together.  Heck, why not let them take up a whole page?

samuelklee replied:
Also, perhaps include the $D$ recovered by ARD somewhere in the figures or captions, since you mention that here?

mbabadi replied:
In this case, the number of samples is not large enough for ARD to detect $D$ reliably.

mbabadi replied:
I put the figures together and made it as large as possible. I need to redo these figures though (the horizontal axis ticks are jumbled).

davidbenjamin commented:
This is great!  Back to @mbabadi

samuelklee commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -670,125 +712,170 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
 
+\item Make a universal panel of normals for both sexes by taking into account targets on sex chromosomes on par with autosomal targets.
+
+\item A reliable method for detecting and excluding outlier samples from the panel of normals, in particular, those with abnormal ploidy.
```
nitpicking, but perhaps rephrase this as an action to be consistent with the other goals listed ("connect", "take", "choose", "use", "make")

mbabadi replied:
Good idea!

samuelklee commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -670,125 +712,170 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
-Suppose we have vectors of read counts over a set of $T$ targets for $S$ samples, $\vn_s, \, s = 1 \ldots S$ where $n_{st}$ is the coverage of sample $s$ at target $t$.  We imagine that laboratory conditions for a particular sample yielding an underlying bias vector $\vb_s$, where $e^{b_{st}}$ is the propensity of target $t$ to be captured, sequenced, and mapped in the preparation of sample $s$.  ...
+\subsubsection{The model}
+
+Suppose we have vectors of read counts over a set of $T$ targets for $S$ samples, $\vn_s, \, s = 1 \ldots S$ where $n_{st}$ is the coverage of sample $s$ at target $t$. In order to include both sexes on an equal footing, we further define a ``ploidy matrix'' $\PP_{st}$ such that $\PP_{st}$ is the ploidy\footnote{For human autosomal targets, $\PP_{st} = 2$ for both sexes. In female samples, $P_{st ...
```
We imagine that laboratory conditions...yield an underlying bias

samuelklee replied:
Also, not sure if we should still refer to $c_{st}$ as the copy number now that the ploidy is accounted for by a separate factor.

samuelklee commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -670,125 +712,170 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
-%
-We can achieve many of the goals listed above by performing probabilistic PCA not on $\vn$ directly but rather on $\vb$.  This is one reason why we choose a Poisson parameter proportional to $e^b$ and not simply $b$ -- probabilistic PCA involves a normal distribution on $\vb_s$, whereas the Poisson parameter must be positive.  Furthermore, since PCA is a linear model with additive effects of diff ...
+
+We can achieve many of the goals listed above by performing probabilistic PCA not on directly $\vn$, but rather on $\vb$. One one hand, the Poisson parameters must be positive and therefore, $\exp(\vb)$ is a well-defined parametrization of the multiplicative bias. On the other hand, a Gaussian model for $\vb$ implies a log-normal distribution for $\exp(\vb)$ which is indeed the expected distribut ...
```
...performing probabilistic PCA not directly on...

On one hand...

I don't know if it's necessary to make the CLT argument.  It's plausible that the underlying assumptions ($N_B \gg 1$, variances of $B_j$ are finite) may not hold.

mbabadi replied:
Do we have a better reason for justifying the lognormal assumption? at least this an argument that it is asymptotically reasonable.

samuelklee commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -670,125 +712,170 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
+\center
+\includegraphics[scale=0.7]{figs/{gauss_poisson_0.1}.pdf}
+\includegraphics[scale=0.7]{figs/{gauss_poisson_10}.pdf}
+\caption{Gaussian approximation to the Poisson likelihood (see Eq.~\ref{eq:gaussian_approx_coverage}). The left and right panels show $\mathrm{Poisson}(n|\alpha\,e^b)$ and $n^{-1}\mathcal{N}(b|\ln(n/\alpha), n^{-1})$, respectively for $\alpha=0.1$ (top) and $\alpha=10.0$ (bottom). The black lines show $b = \ln(n/\alpha)$ the maximum likelihood bias estimate. The Gaussian approximation breaks down ...
```
Might also be nice to plot the ratio or the relative error between the two quantities.

samuelklee replied:
Also, perhaps shift the x-axis labels to be between the ticks?  Same with the solid line?

mbabadi replied:
I agree that the plot could be more informative (i.e. plotting relative error). I don't get the comment regarding the x-axis labels -- can you elaborate?

samuelklee commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -670,125 +712,170 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
-The approximation that led to this can be interpreted as the replacement $\vSigma_s \rightarrow \bar{\vSigma} \equiv \frac{1}{S} \sum_s \vSigma_s$ and we will consider it again below when we discuss EM updates of $\vW$ and $\vPsi$.
+%
+To simplify the discussion, let us assume that all samples share the same targets, i.e. $\vM_s = \mathbb{I}$.
+First, note that $\vPsi + \vW \vW^T$ is the covariance of log-coverage due to experimental conditions, both those included explicitly in $\vW$ and the residual, unexplained covariance $\vPsi$. The quantities $\vPsi + \vSigma_s + \vW \vW^T$ are essentially weighting factors that decrease the role of lower-coverage samples (i.e. those with larger $\vSigma$) in the likelihood. If variations in $\vSi ...
```
Considering the next paragraph, I think this reads slightly better with "...a typical sample-independent value $\bar{\Sigma}_t$ is a very reasonable approximation."  (note also typos "indendent" and "approxmiation" were corrected)

samuelklee commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -670,125 +712,170 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
+First, note that $\vPsi + \vW \vW^T$ is the covariance of log-coverage due to experimental conditions, both those included explicitly in $\vW$ and the residual, unexplained covariance $\vPsi$. The quantities $\vPsi + \vSigma_s + \vW \vW^T$ are essentially weighting factors that decrease the role of lower-coverage samples (i.e. those with larger $\vSigma$) in the likelihood. If variations in $\vSi ...
 
-For fixed $\vc$ we may obtain maximum likelihood estimates of $\vW$ and $\vSigma$ by an iterative EM approach in which we alternate between computing the Gaussian posteriors of each $\vz_s$ (E step) and maximizing the log-likelihood with respect to $\vW$, $\vSigma$, and $\vm$ (M step).  This is similar to the factor analysis model discussed in Chapter 12 of Bishop.  We note calling $\vc$ at each  ...
+A natural choice for $\bar{\Sigma}_t$ will be made apparent soon. In the meantime, we notice that the $\vSigma_s$ always appears in addition to $\vPsi$ such that the specific choice of $\bar{\Sigma}_t$ may seem to be immaterial: if $\vPsi^*$ is the MLE estimate obtained using the choice $\bar{\vSigma}^*$, we may infer that $\vPsi^* + \bar{\vSigma}^* - \bar{\vSigma}'$ is the MLE estimate had we us ...
```
...we notice that $\vSigma_s$ always appears...

mbabadi replied:
fixed

samuelklee commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -670,125 +712,170 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
-The E step follows from substituting $\vPsi \rightarrow \vPsi_s \equiv  \vPsi + \vSigma_s$ in Bishop's Equations 12.66-67 to get
-%
+
+\subsubsection{EM algorithm for obtaing MLE of $(\vW, \vPsi, \vm)$:}
```
obtaing -> obtaining

samuelklee replied:
left off reading here, will continue later!

mbabadi replied:
fixed.

samuelklee commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -670,125 +712,170 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
 %
-
-If we do not use the approximation for $\vm$, its M step is
+The stationarity condition for $\mathcal{L}$ with respect to $\vm$ gives:
```
Not sure if "stationarity" is the correct terminology here?

mbabadi replied:
Stationarity = derivative is zero, right?

samuelklee replied:
i think "stationary" is more common in this context?

samuelklee commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -829,44 +916,99 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
+\begin{equation}
+R_f \equiv \exp\left(-\frac{\lambda}{2} \sum_{s=1}^S\kappa_s\right) = \exp\left(-\frac{\lambda}{T}\sum_{s=1}^S\sum_{t,t'=1}^{T-1}f(t-t') \, W_{t\mu} \, W_{t'\nu} \, z_{s\mu} \, z_{s\nu}\right).
+\end{equation}
+We will dicuss a proper choice of $\lambda$ later. Since this regularizer is quadratic in $z$, the Gaussian structure of the likelihood is preserved and the E step remains as simple as before. The only difference is the presence of an additional term in $\vG_s^{-1}$:
```
dicuss -> discuss

mbabadi replied:
fixed

mbabadi replied:
fixed.

samuelklee commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -829,44 +916,99 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
+\begin{equation}
+\vG_s^{-1} \rightarrow \vI + \vW^T \left[\vM_s \vPsi_s^{-1} + \lambda \vF\right]\vW.
+\end{equation}
+Since $\vF$ is not diagonal in the target space, a naive matrix multiplication implies a multiplication complexity of $\mathcal{O}(D^2 T^2)$ for the new term. However, this complexity can be reduced to a manageble $\mathcal{O}(D^2 T \log T)$ using FFT:
```
manageble -> manageable

mbabadi replied:
fixed.

mbabadi replied:
fixed.

samuelklee commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -829,44 +916,99 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
+\center
+\includegraphics[scale=0.4]{figs/comp_random_events_10.pdf}
+\includegraphics[scale=0.4]{figs/comp_random_events_20.pdf}
+\caption{Comparison of PCA with the probablistic coverage model in different modes for uncorrelated events. (top) $D=10$, (bottom) $D=20$.}
```
probablistic -> probabilistic

mbabadi replied:
fixed.

samuelklee commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -829,44 +916,99 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
+\center
+\includegraphics[scale=0.4]{figs/comp_corr_events_10.pdf}
+\includegraphics[scale=0.4]{figs/comp_corr_events_20.pdf}
+\caption{Comparison of PCA with the probablistic coverage model in different modes for correlated events. (top) $D=10$, (bottom) $D=20$.}
```
probablistic -> probabilistic

mbabadi replied:
fixed.

samuelklee commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -670,125 +712,170 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
 %
 \begin{align}
-P(\vW) =& \prod_k \left( \frac{\alpha_\mu}{2 \pi} \right)^{T/2} \exp \left( -\frac{1}{2} \alpha_\mu \sum_t W_{t\mu}^2  \right).
+P(\vW) =& \prod_\mu \left( \frac{\alpha_\mu}{2 \pi} \right)^{T/2} \exp \left( -\frac{1}{2} \alpha_\mu \sum_t W_{t\mu}^2  \right).
```
Nitpicking, but why mix Greek and Latin indices?

mbabadi replied:
Greek indices are used for latent space indices.

mbabadi replied:
Greek letters are for latent space indices, and $\alpha$ belongs to the latent space.

samuelklee replied:
ah ok!  i'd mention this explicitly in the text if you feel it is useful.

samuelklee commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -829,44 +916,99 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
+\noindent {\bf Why does it matter? ---} We have no theoretical guarantee that the MLE problem for $(\vW, \vm, \vPsi, \vc)$ is convex. In all likelihood, if one of the samples has a large germline CNV event, it may be picked up as a principal component and get washed out such that the MLE for $\vc$ fails to call that CNV event. It is possible that the likelihood function has numerous local maxima  ...
 
-In exact mode the leading cost is a few terms of $O(D^2 T S)$ flops, each with small prefactors, per iteration.  In approximate mode it is $O(D T S)$ flops.  Assuming a total prefactor of $10$ and $T = 2 \times 10^5$, $D = 10$, $S = 500$ a full EM iteration costs $10^{11}$ flops in exact mode and $10^{10}$ in approximate mode.  On a single 1 GHz core ($10^9$ flops per second) this comes out to ro ...
+\noindent {\bf (Idea 1) Blind source separation ---} One remedy is to use a blind source separation approach, such as independent component analysis (ICA), to separate the signal from the noise as a first step, followed by learning the latent features of the noise using PCA. In ICA-like methods, one decomposes the signal into additive subcomponents and minimizes the mutual information between the ...
```
non-Gaussanity -> non-Gaussianity

mbabadi replied:
fixed.

samuelklee commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -829,44 +916,99 @@ \subsection{Generative Model for Read Counts} \label{coverage-model}
+\end{array}
+\right.
+\end{equation}
+Here, $k^* = \min(k, T - 1 - k)$, $k_l \sim \lfloor T / \ell_\mathrm{max} \rfloor$ and $k_l \sim \lfloor T / \ell_\mathrm{min} \rfloor$, where $\ell_\mathrm{min}$ and $\ell_\mathrm{max}$ denote roughly the minimum and maximum length of the CNV events in the units of targets. The filtered spectral power of the noise in sample $s$ is given as:
```
I think it's most likely fine to work in target-index space, but am curious about how target spacing might affect things.

mbabadi replied:
In principle, the best strategy is to (1) make a new matrix $W_{t\mu} \rightarrow W_{x\mu}$ with $x$ representing the genomic position, and $W_{x\mu} \equiv W_{t(x)\mu}$ where $t(x)$ is the target that overlaps with $x$, and (2) perform Fourier filtering on $W_{x\mu}$. If target spacing is not too non-uniform, this might be unnecessarily extra work. This is easy to code but lets to -extremely- large matrices and time-consuming FFT operations. If the target spacing is not terribly non-uniform, we may leave it as it is. What do you guys think? @davidbenjamin @samuelklee

davidbenjamin replied:
I think it's probably close enough to uniform, especially since we care about long length scales at which individual variations in target spacing will wash out.

mbabadi replied:
In principle, the best strategy is to (1) make a new matrix $W_{t\mu} \rightarrow W_{x\mu}$ with $x$ representing the genomic position, and $W_{x\mu} \equiv W_{t(x)\mu}$ where $t(x)$ is the target that overlaps with $x$, and (2) perform Fourier filtering on $W_{x\mu}$. If target spacing is not too non-uniform, this might be unnecessarily extra work. This is easy to code but lets to -extremely- large matrices and time-consuming FFT operations. If the target spacing is not terribly non-uniform, we may leave it as it is. What do you guys think? @davidbenjamin @samuelklee

samuelklee commented:
done with my review.  nice work @mbabadi, looks awesome!

mbabadi commented:
OK guys @davidbenjamin @samuelklee , thanks for your in-depth review. The typos that you diligently caught were due to not enabling spell checker on Sublime Text :) I implemented all of the changes. The figures still need to be updated, and I will do it soon.

In the meantime, I have added to very short sections: Sec. 2 in pg. 18 on estimating read depth, and Sec. C pg. 25 on inferring sex genotypes. They should be pretty easy to review -- @davidbenjamin , could you please take a look?

davidbenjamin commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -773,12 +773,31 @@ \subsubsection{The model}
 A natural choice for $\bar{\Sigma}_t$ will be made apparent soon. In the meantime, we notice that the $\vSigma_s$ always appears in addition to $\vPsi$ such that the specific choice of $\bar{\Sigma}_t$ may seem to be immaterial: if $\vPsi^*$ is the MLE estimate obtained using the choice $\bar{\vSigma}^*$, we may infer that $\vPsi^* + \bar{\vSigma}^* - \bar{\vSigma}'$ is the MLE estimate had we us ...
 
+\subsubsection{Estimating the mean read depth $d_s$}\label{sec:read_depth_estimation}
+Let us define $\lambda_{st} \equiv \PP_{st} c_{st} e^{b_{st}}$ as the total multiplicative bias per sample per target. Assuming that $\lambda_{st}$ is sharply peaked around its mode, the MLE for $d_s$ can be easily found by approximating the Poisson distributions with Gaussian like before:
```
Gaussian -> Gaussians

davidbenjamin replied:
I would prefer to reserve "bias" for `e^{b_{st}}` alone and refer to this broader quantity as "total multiplicative factor".

mbabadi replied:
Done

davidbenjamin commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -773,12 +773,31 @@ \subsubsection{The model}
+P(\vn_s | d_s \boldsymbol{\lambda}_s) &= \prod_{t|M_{st} \neq 0} \mathrm{Poisson}(n_{st} | d_s \lambda_{st}) \simeq \prod_{t|M_{st} \neq 0} \norm(n_{st}|d_s \lambda_{st}, d_s \lambda_{st})\nonumber\\
+&\propto \prod_t \left[\frac{1}{\sqrt{d_s \lambda_{st}}}\exp\left(-\frac{(n_{st} - d_s \lambda_{st})^2}{2 d_s \lambda_{st}}\right)\right]^{M_{st}}.
+\end{align}
+Replacing $\lambda_{st}$ with its mode, the log likelihood of read depth is given as:
```
I'm not sure what you mean by this since `\lambda_{st}` is still in the following equation.  Do you mean "maximizing wrt `d_s` and dropping terms containing only `\lambda_{st}`"?

mbabadi replied:
We need to integrate out lambda to get the likelihood for d. In subsequent
eqs, it is assumed that lambda is set to its mode. I can state that
explicitly.

On Jul 26, 2016 9:48 AM, "David Benjamin" <notifications@github.com> wrote:

In docs/CNVs/CNV-methods.tex
<https://github.com/broadinstitute/gatk-protected/pull/556#discussion_r72251909>
:

>
>  A natural choice for $\bar{\Sigma}_t$ will be made apparent soon. In the meantime, we notice that the $\vSigma_s$ always appears in addition to $\vPsi$ such that the specific choice of $\bar{\Sigma}_t$ may seem to be immaterial: if $\vPsi^*$ is the MLE estimate obtained using the choice $\bar{\vSigma}^*$, we may infer that $\vPsi^* + \bar{\vSigma}^* - \bar{\vSigma}'$ is the MLE estimate had we used $\bar{\vSigma}'$ instead. However, the constraint $\Psi_t > 0$ implies that $\Psi^*_t + \bar{\Sigma}_t > 0$, such that our choice of $\bar{\Sigma}_t$ effectively sets a lower bound on the unexplained variance.\\
>
> +\subsubsection{Estimating the mean read depth $d_s$}\label{sec:read_depth_
estimation}
> +Let us define $\lambda_{st} \equiv \PP_{st} c_{st} e^{b_{st}}$ as the total multiplicative bias per sample per target. Assuming that $\lambda_{st}$ is sharply peaked around its mode, the MLE for $d_s$ can be easily found by approximating the Poisson distributions with Gaussian like before:
> +\begin{align}
> +P(\vn_s | d_s \boldsymbol{\lambda}_s) &= \prod_{t|M_{st} \neq 0} \mathrm{Poisson}(n_{st} | d_s \lambda_{st}) \simeq \prod_{t|M_{st} \neq 0} \norm(n_{st}|d_s \lambda_{st}, d_s \lambda_{st})\nonumber\\
> +&\propto \prod_t \left[\frac{1}{\sqrt{d_s \lambda_{st}}}\exp\left(-\frac{(n_{st} - d_s \lambda_{st})^2}{2 d_s \lambda_{st}}\right)\right]^{M_{st}}.
> +\end{align}
> +Replacing $\lambda_{st}$ with its mode, the log likelihood of read depth is given as:

I'm not sure what you mean by this since \lambda_{st} is still in the
following equation. Do you mean "maximizing wrt d_s and dropping terms
containing only \lambda_{st}"?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/broadinstitute/gatk-protected/pull/556/files/7793eed22caa2c3ea1dc3dd8174fce8a8a7051cf..046affe48123b237c5c17a70944bfa3422cdfff3#r72251909>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AOmMjexhUokGaz9wza4rXNaJ16WEJp4Jks5qZhArgaJpZM4Iz5Lx>
.

davidbenjamin commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -773,12 +773,31 @@ \subsubsection{The model}
+\end{equation}
+Maximizing yields a quadratic equation with one acceptable solution:
+\begin{equation}
+d_s = \frac{\sqrt{4\llangle \vn^2 \rrangle_s \llangle \boldsymbol{\lambda}^2 \rrangle_s + \llangle \boldsymbol{\lambda} \rrangle_s^2} - \llangle \boldsymbol{\lambda} \rrangle_s}{2 \llangle \boldsymbol{\lambda}^2 \rrangle_s},
```
Maximizing Eq 61 wrt `d` I get something slightly different:
[sqrt( 1 + 4 <<lambda>><<n^2/lambda>>) - 1]/[2 <<lambda>>].  Your result follows if you multiply throughout by `\lambda` in an intermediate step, but I believe this can't be done since `\lambda` depends on `t`

mbabadi replied:
Huh good catch! I got lazy and dropped "st" labels in my scribblings, and it "allowed" me to multiply the sides by "\lambda".

davidbenjamin commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -773,12 +773,31 @@ \subsubsection{The model}
+\begin{equation}
+\llangle \mathbf{v} \rrangle_s \equiv \frac{\sum_t M_{st} v_{st}}{\sum_t M_{st}}.
+\end{equation}
+If $n_{st} \gg 1$, we may approximate the result as $d_s \simeq \left(\llangle \vn^2 \rrangle_s / \llangle \boldsymbol{\lambda}^2 \rrangle_s\right)^{\frac{1}{2}}$. Before learning the parameters of the coverage model, we may set $\lambda_{st} \rightarrow \PP_{st}$ to obtain a first estimate of $d_s$. This estimate can be updated along the way as better estimates for $c_{st}$ and $b_{st}$ is found ...
```
are found

davidbenjamin commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -1010,6 +1029,50 @@ \subsubsection{Results}
 
 
+\subsection{Inferring sex genotype from raw read counts}
+The probabilistic target coverage model requires the germline copy number of the targets as an input for dealing with cohorts composed of samples with different sexes. If the phenotypic sex of the human subject associated to a sample is known, the nominal germline ploidy of each target is readily known. Unfortunately, the gender annotation is not available for most samples. In this section, we pr ...
```
How about "for dealing with mixed-sex cohorts"?

davidbenjamin replied:
I would delete "If the phenotypic sex. . . not available for most samples" and change the last sentence of this paragraph to "In this section we propose a simple statistical test for inferring sex from raw read counts when the phenotypic sex is not known."

mbabadi replied:
Agreed.

davidbenjamin commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -1010,6 +1029,50 @@ \subsubsection{Results}
+\end{array}
+\right.
+\end{equation}
+where $\rho$ is {\em average read count per base per strand}, $l_t$ is the target length (number of overlapping bases), and $\epss_M \sim 10^{-5}$ is the (small) probability of a mapping error that may result in a (small) number of reads to be mapped to the Y chromosome for an XX genotype sample. Similarly, for the YY genotype we have:
```
"Strand" means half of a double helix.  Do you mean "homologous copy"?

davidbenjamin replied:
Where did you get this mapping error rate from?  Does it include the Y chromosome's pseudo-autosomal region?

mbabadi replied:
What I mentioned (1e-5) was a typical number -- $\epss_M$ comes from mapping quality. Once we get from BAM files to ReadCountCollection, mapping quality information is lost. In practice, a ballpark estimate is sufficient though: XX and XY samples are sufficiently easy to tell apart. That being said, if we set $\epss_M = 0$, we always get XY genotype since all XX samples have a small number of their reads erroneously aligned to Y chr. I am not sure if I understand the role of pseudo-autosomal regions -- do you mean somehow we expect a higher mapping error rate, compared to autosomal targets, in the pseudo-autosomal regions of X and Y chromosomes?

davidbenjamin commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -1010,6 +1029,50 @@ \subsubsection{Results}
+\end{align}
+We have defined $L_{X(Y)} \equiv \sum_{t \in T_{X(Y)}} l_t$ as the total length of targets in X and Y chromosomes.\\
+
+Having a likelihood function, the logarithmic odds of $XX$ to $XY$ genotype is readily found from the Bayes theorem:
```
"the Bayes theorem" -> "Bayes' theorem"

davidbenjamin commented on docs/CNVs/CNV-methods.tex:
```diff
@@ -1010,6 +1029,50 @@ \subsubsection{Results}
+\begin{equation}
+\log\,\mathrm{odds}(\text{XX}/\text{XY}) = \log P(\vn|\text{XX}) + \log P(\text{XX}) - \log P(\vn|\text{XY}) - \log P(\text{XY}).
+\end{equation}
+If the compounds reads likelihood model is used, the compound likelihood functions must be accordingly:
```
compounds -> compound

davidbenjamin commented:
@mbabadi nice work; done with review.

samuelklee commented:
@mbabadi These changes were included in the other branch, right?  If so, can we close?

mbabadi commented:
@samuelklee yes, closed.
